### PR TITLE
Use List of Connections in CandidateRoute

### PR DESCRIPTION
**Pull request by kschrab**
## Type of this PR 

- [x] Bug fix
- [x] Enhancement

## Description

Previously, CandidateRoute held a list of node Ids which belong to a route. Based on those nodes, a list of connections had been extracted to be used in SUMO. However, this behavior breaks as soon as there are parallel connections with the same start and end node and no nodes in between. Also, the code to generate the list of connections from nodes was very hard to understand.

Now, the CandidateRoute holds a list of connection Ids directly extracted from the edges of its GraphHopper route. Any code which tried to generate the list of connections from node Ids could be eliminated. All tests have been fixed accordingly, now checking for connection Ids instead of node Ids.

## Issue(s) related to this PR

 * Internal Issue 156 
 
## Affected parts of the online documentation

No

## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All commits are signed-off (`-s`) with your mail address of your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [ ] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been observed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

**Comment by kschrab** [comment]
@schwepmo please review :)

**Review by schwepmo — CHANGES_REQUESTED** [review]
Looks good! I made a few small comments.

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/traci/writer/TrafficLightPhaseWriter.java [hunk 1]
```diff
@@ -52,9 +52,9 @@ public void writeVariableArgument(DataOutputStream out, SumoTrafficLightLogic.Ph
         stringWriter.writeVariableArgument(out, argument.getPhaseDef());
 
         out.writeByte(TraciDatatypes.DOUBLE);
-        out.writeDouble(argument.getDuration() / 1000d);
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/traci/writer/TrafficLightPhaseWriter.java:55]: correct :)

File: lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java [hunk 2]
```diff
@@ -268,166 +275,18 @@ private CandidateRoute preparePath(QueryGraph queryGraph, Path newPath, Node pre
...
-                }
-                i++;
-            }
-
-            if (!(indexOnConnectionNode1 >= 0 && indexOnConnectionNode2 > indexOnConnectionNode1)) {
-                return false;
-            }
+            return source.getConnectionId().equals(route.getConnectionIds().get(0));
         }
         return true;
```
[the remark this answers is not in the document]
Comment (kschrab) [reply lib/mosaic-routing/src/main/java/org/eclipse/mosaic/lib/routing/graphhopper/GraphHopperRouting.java:289]: Yes, if there's no explicit connection given in the source position, than it does not matter which the first connection actually is.

File: fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/navigation/RoadPositionFactory.java [hunk 3]
```diff
@@ -103,21 +104,19 @@ public static IRoadPosition createAlongRoute(
...
+     * @param currentRouteConnectionIds The list of connectionIds, that the {@link IRoadPosition} should be created for.
+     * @param laneIndex                 The lane index.
      * @return An {@link IRoadPosition} of the last edge in route.
      */
-    public static IRoadPosition createAtEndOfRoute(List<String> currentRouteNodeIds, int laneIndex) {
-        final String startNodeOfLastEdge = currentRouteNodeIds.get(currentRouteNodeIds.size() - 2);
-        final String endNodeOfLastEdge = currentRouteNodeIds.get(currentRouteNodeIds.size() - 1);
-        return refine(new SimpleRoadPosition(startNodeOfLastEdge, endNodeOfLastEdge, laneIndex, -1d));
+    public static IRoadPosition createAtEndOfRoute(List<String> currentRouteConnectionIds, int laneIndex) {
+        return refine(new SimpleRoadPosition(Iterables.getLast(currentRouteConnectionIds), laneIndex, -1d, 0d));
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/navigation/RoadPositionFactory.java:121]: Done

**Review by rprotzmann — APPROVED** [review]

**Pull request merged**